### PR TITLE
Fixed building on go1.6, go1.7 and go1.8

### PR DESCRIPTION
**Pull request by xaionaro**
Relates: https://github.com/trezor/trezord-go/issues/80

The most disputable part of the PR is related to changes in file server/http.go: https://github.com/trezor/trezord-go/pull/81/files#diff-52d328bdc6be1ea07f80fdb67869a272

### CODE REVIEW

**Comment by xaionaro** [comment]
@karel-3d here's the pull request :)

**Comment by xaionaro** [comment]
It was not enough time to also fix on go1.5. The error on go1.5 is:
```
# github.com/trezor/trezord-go/usb/lowlevel
malformed DWARF TagVariable entry
```

File: server/http.go [hunk 1]
```diff
@@ -77,9 +83,5 @@ func (s *Server) logRequest(handler http.Handler) http.Handler {
 }
 
 func (s *Server) Run() error {
-	return s.https.ListenAndServe()
-}
-
-func (s *Server) Close() error {
-	return s.https.Close()
+	return s.ListenAndServe()
```
Comment (karel-3d) [inline server/http.go]: why was this needed, with serverPrivate?

Comment (karel-3d) [reply server/http.go]: ...also why is Close removed?

File: vendor/gopkg.in/natefinch/lumberjack.v2/linux_test.go [hunk 2]
```diff
@@ -123,7 +123,7 @@ func TestCompressMaintainMode(t *testing.T) {
 	filename2 := backupFile(dir)
 	info, err := os.Stat(filename)
 	isNil(err, t)
-	info2, err := os.Stat(filename2+compressSuffix)
+	info2, err := os.Stat(filename2 + compressSuffix)
```
Comment (karel-3d) [inline vendor/gopkg.in/natefinch/lumberjack.v2/linux_test.go]: not sure if it's good idea to reformat vendored packages?

Comment (xaionaro) [reply vendor/gopkg.in/natefinch/lumberjack.v2/linux_test.go]: Sorry. Didn't notice it was in `vendor/`. Reverted.

File: server/http.go (hunk 1 above)
Comment (karel-3d) [reply server/http.go]: Oh I see in the commit comment, sorry

**Comment by karel-3d** [comment]
This looks fine. Thanks a lot

**Comment by karel-3d** [comment]
`malformed DWARF TagVariable entry` has something to do with the cgo magic

File: server/http.go (hunk 1 above)
Comment (xaionaro) [reply server/http.go]: There's no `(*http.Server).Close()` function in go1.7. I noticed that `(*server.Server).Close()` is not used anywhere in `trezord-go`. And didn't know do you need this `Close()` method or not, so I developed a solution that preserves it only for >=go1.8 and removes for <go1.8

To do that I just used unnamed embedded structure (to passthrough all methods of `*http.Server` to `server.Server`). If `*http.Server` has implemented `Close()` (>=go1.8) it will be passthrough'ed to `server.Server`. If not then won't.

But if I use directly `*http.Server` within `Server` it won't be public while `https` was public. So to preserve this property I made an intermediate structure `serverPrivate` to make `*http.Server` to be private.

So `Close()` is removed in this patch because it's already implemented in `*http.Server` and passthrough'ed to `Server`.

Comment (karel-3d) [reply server/http.go]: Ah, the plan was to use Close() in defer in main, but we forgot to do that :D thanks for indirectly noticing

Comment (xaionaro) [reply server/http.go]: If you need `Close()` I can prepare a new Pull Request to fix the building (after you add `Close()` into `main`).

Comment (karel-3d) [reply server/http.go]: Hm, but now I cannot add `Close()` to be deferred in main, since I don't know whether it is or isn't go1.7, right. There are some ways I can find to do that, but they all seem overly complex.

I guess this is fine for now since exiting the app stops the server anyway.

Comment (xaionaro) [reply server/http.go]: > Hm, but now I cannot add Close() to be deferred in main, since I don't know whether it is or isn't go1.7, right

You can add file (let's call it for example `http_go1.7.go`) with a content like (writing in blind, didn't check this):
```
// +build !go1.8
func (s *Server) Close() error {
    return s.Shutdown()
}
```

Or a dirty hack :) :
```
type serverCloser interface {
    Close() error
}
func (s *Server) Close() error {
    closer, ok := (interface{})((*http.Server)(s)).(serverCloser)
    if ok {
        return closer.Close()
    }
    return s.Shutdown()
}
```

Comment (karel-3d) [reply server/http.go]: Ok that all looks too hacky :) this is fine for now, since it is not a regression anyway, the server closes on exit fine

Comment (xaionaro) [reply server/http.go]: In my very humble opinion, this looks nice — `server/http_go-lt1.8.go`:
```go
// +build !go1.8

func (s *Server) Close() error {
    return s.Shutdown()
}
```

Anyway. Thanks again for the merging)

Comment (karel-3d) [reply server/http.go]: Shutdown() doesn't seem to work on 1.7 either

**Comment by karel-3d** [comment]
Btw I tried to build with go 1.5 and it worked fine. (I used binary tarball go from their official website.)

**Comment by karel-3d** [comment]
Travis thinks the same 

https://travis-ci.org/trezor/trezord-go/jobs/402220125

File: server/http.go (hunk 1 above)
Comment (xaionaro) [reply server/http.go]: Ok. I'll research for another way.

**Comment by xaionaro** [comment]
> Btw I tried to build with go 1.5 and it worked fine. (I used binary tarball go from their official website.)

Khm. That's odd.

```
$ GIMME_OUTPUT="$(gimme 1.5)" && eval "$GIMME_OUTPUT"
go version go1.5 linux/amd64
$ rm -rf go
$ go get github.com/trezor/trezord-go
# github.com/trezor/trezord-go/usb/lowlevel
malformed DWARF TagVariable entry
```

File: server/http.go (hunk 1 above)
Comment (xaionaro) [reply server/http.go]: Here's an implementation of Close(): https://github.com/trezor/trezord-go/pull/82

File: core/core.go [hunk 3]
```diff
@@ -150,14 +162,12 @@ func (c *Core) createEnumerateEntries(infos []USBInfo) []EnumerateEntry {
 		}
 		entries = append(entries, e)
 	}
-	sortEntries(entries)
+	entries.Sort()
```
[the remark this answers is not in the document]
Comment (xaionaro) [reply core/core.go]: I disagree. Sort should be called via internal method (not via a global special function). It's internal stuff of EnumerateEntries and there's no need to do a function outside of this entity. But this's just my *subjective* sense of beauty and code-auto-documentation :)
